src/com/facebook/buck/jvm/java/JavaDocJar.java

```diff
@@ -0,0 +1,135 @@
+import com.facebook.buck.rules.AddToRuleKey;
+import com.facebook.buck.rules.BuildContext;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildTargetSourcePath;
+import com.facebook.buck.rules.BuildableContext;
+import com.facebook.buck.rules.SourcePath;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.step.Step;
+import com.facebook.buck.step.fs.MakeCleanDirectoryStep;
+import com.facebook.buck.step.fs.MkdirStep;
+import com.facebook.buck.step.fs.RmStep;
+import com.facebook.buck.zip.ZipStep;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+
+
+public class JavaDocJar extends AbstractBuildRule implements HasMavenCoordinates {
+
+  @AddToRuleKey
+  private final ImmutableSortedSet<SourcePath> sources;
+  private final Path output;
+  private final Path temp;
+  private final Optional<String> mavenCoords;
+
+  public JavaDocJar(
+      BuildRuleParams params,
+      SourcePathResolver resolver,
+      ImmutableSortedSet<SourcePath> sources,
+      Optional<String> mavenCoords) {
+    super(params, resolver);
+    this.sources = sources;
+    BuildTarget target = params.getBuildTarget();
+    this.output =
+        BuildTargets.getGenPath(
+            getProjectFilesystem(),
+            target,
+            String.format("%%s%s", Javac.JAVADOC_JAR));
+    this.temp = BuildTargets.getScratchPath(getProjectFilesystem(), target, "%s-javadoc");
+    this.mavenCoords = mavenCoords;
+  }
+
+  @Override
+  public ImmutableList<Step> getBuildSteps(
+      BuildContext context, BuildableContext buildableContext) {
+    ImmutableList.Builder<Step> steps = ImmutableList.builder();
+
+    Path javadocOutput = new File(output.getParent().toFile(), "javadoc").toPath();
+    steps.add(new RmStep(getProjectFilesystem(), javadocOutput, /* force deletion */ true, true));
+    steps.add(new RmStep(getProjectFilesystem(), output, true));
+    steps.add(new MkdirStep(getProjectFilesystem(), javadocOutput));
+    steps.add(new MakeCleanDirectoryStep(getProjectFilesystem(), temp));
+
+
+    List<String> javaDocArgs = new LinkedList<>();
+    javaDocArgs.addAll(Arrays.asList("-notimestamp", "-private", "-subpackages", ".",
+        "-d", javadocOutput.toString()));
+
+    StringBuilder classpath = new StringBuilder(".");
+    String classpathSeparator = ":";
+    if (System.getProperty("os.name").startsWith("Windows")) {
+      classpathSeparator = ";";
+    }
+    for (BuildRule dep : getDeps()) {
+      BuildTargetSourcePath abiJar = new BuildTargetSourcePath(dep.getBuildTarget().withAppendedFlavors(CalculateAbi.FLAVOR));
+      if (abiJar.getResolvedPath().isPresent()) {
```

Comment:
These two lines just create a new `BuildTarget` and then basically ask whether the (unset) second parameter of the `BuildTargetSourcePath` is present. They're a highly elaborate no-op :)

Better would be something like:

```java
ImmutableSortedSet.Builder<Path> allJars = ImmutableSortedSet.builder();
for (BuildRule dep : getDeps()) {
  if (dep instanceof HasClasspathEntries) {
    allJars.add(((HasClasspathEntries) dep).getTransitiveClasspathEntries().values());
  }
}
```

You can then use the `FluentIterable` to first transform everything into a `String` (you'll find `Functions.toStringFunction()` helpful) and then use `Joiner` to mash the transformed set together.

I like to keep the class path ordered to make it easier to read. You may need to resolve the absolute path to each output.